### PR TITLE
Proposer Server Filters Attestation by Canonical

### PR DESCRIPTION
**Pull request by terenc3t**
This is part of #2268

Check whether the attestation is voting on the canonical blocks, if not, don't include them in the proposed block.

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -206,3 +208,12 @@ func (c *ChainService) ChainHeadRoot() ([32]byte, error) {
 	}
 	return root, nil
 }
+
+// IsCanonical returns true if the input block hash of the corresponding slot
+// is part of the canonical chain. False otherwise.
+func (c *ChainService) IsCanonical(slot uint64, hash []byte) bool {
+	if canonicalHash, ok := c.canonicalBlocks[slot]; ok {
```
Comment (rauljordan) [inline beacon-chain/blockchain/service.go]: RMutex here?

File: beacon-chain/blockchain/fork_choice.go [hunk 2]
```diff
@@ -142,6 +142,12 @@ func (c *ChainService) ApplyForkChoiceRule(
 	if err != nil {
 		return fmt.Errorf("could not run fork choice: %v", err)
 	}
+	headRoot, err := hashutil.HashBeaconBlock(head)
+	if err != nil {
+		return fmt.Errorf("could not hash head block: %v", err)
+	}
+	c.canonicalBlocks[head.Slot] = headRoot[:]
```
Comment (rauljordan) [inline beacon-chain/blockchain/fork_choice.go]: Use mutexes for updating/deleting

File: beacon-chain/blockchain/service.go [hunk 3]
```diff
@@ -206,3 +211,22 @@ func (c *ChainService) ChainHeadRoot() ([32]byte, error) {
...
+	defer c.canonicalBlocksLock.RUnlock()
+	if canonicalHash, ok := c.canonicalBlocks[slot]; ok {
+		return bytes.Equal(canonicalHash, hash)
+	}
+	return false
+}
+
+// InsertsCanonical inserts a canonical block hash to its corresponding slot.
+// This is used for testing purpose.
+func (c *ChainService) InsertsCanonical(slot uint64, hash []byte) {
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/service.go]: Please remove this method. We shouldn't have methods like this unless absolutely necessary. It's also confusing that there is the duplicated logic in fork_choice.go

Comment (terenc3t) [reply beacon-chain/blockchain/service.go]: Will remove in a subsequent PR

**Pull request merged**